specification/provider/providers.md

```diff
@@ -1,5 +1,159 @@
+
+##### Requirement 2.7
+
+> In cases of normal execution, the `provider` **MUST NOT** populate the `flag resolution` structure's `error code` field, or otherwise must populate it with a null or falsy value.
+
+##### Requirement 2.8
+
+> In cases of abnormal execution, the `provider` **MUST** indicate an error using the idioms of the implementation language, with an associated error code having possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+The provider might throw an exception, return an error, or populate the `error code` object on the returned `flag resolution` structure to indicate a problem during flag value resolution.
+
+##### Condition 2.9
+
+> The implementation language supports generics (or an equivalent feature).
+>
+> ##### Conditional Requirement 2.9.1
+>
+> > The `flag resolution` structure **SHOULD** accept a generic argument (or use an equivalent language feature) which indicates the type of the wrapped `value` field.
+
+```
+// example boolean flag value resolution with generic argument
+FlagResolution<boolean> resolveBooleanValue(string flagKey, boolean defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example string flag value resolution with generic argument
+FlagResolution<string> resolveStringValue(string flagKey, string defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example number flag value resolution with generic argument
+FlagResolution<number> resolveNumberValue(string flagKey, number defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+
+// example structure flag value resolution with generic argument
+FlagResolution<MyStruct> resolveStructureValue(string flagKey, MyStruct defaultValue, context: EvaluationContext, options: FlagEvaluationOptions);
+```
+
+#### Context Transformation
+
+Feature flag management systems often define structures representing arbitrary contextual data pertaining to the runtime, user, or application. The context transformer defines a simple interface to transform the OpenFeature `evaluation context` to such a structure, mapping values appropriately.
+
+See [evaluation context](../evaluation-context/evaluation-context.md).
+
+##### Requirement 2.10
+
+> The provider interface **SHOULD** define a `context transformer` method or function, which can be optionally implemented in order to transform the `evaluation context` prior to flag value resolution.
+
+The OpenFeature `client` applies the transformer function before passing the returned value (the `transformed context`) to the provider resolution methods.
+
+```
+class MyProvider implements Provider {
+  //...
+
+  // implementation of context transformer
+  MyProviderContext transformContext(EvaluationContext context) {
+    return new MyProviderContext(context.email, context.ip, context.httpMethod);
+  }
+
+  //...
+}
+```
```

Comment:
See above. I think this is unnecessary as it's implicit that the provider has to transform the content from "what openfeature gives it" into "what the provider's SDK actually needs"

Reply:
See my other comment here: https://github.com/open-feature/spec/pull/41#discussion_r860825935